### PR TITLE
feat: handle ckb unlock, sign and send unlock tx

### PR DESCRIPTION
**Pull request by pygman**

### CODE REVIEW

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 1]
```diff
@@ -898,6 +938,192 @@ export class CkbHandler {
...
+        logger.error(`CKB handleTodoUnlockRecords error:${e.stack}`);
+      },
+    });
+  }
+
+  async todoUnlockRecordsHandler(ownerTypeHash: string, generator: CkbTxGenerator): Promise<void> {
+    if (!TransferOutSwitch.getInstance().getStatus()) {
+      logger.info('TransferOutSwitch is off, skip handleTodoUnlockRecords');
+      return;
+    }
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:965]:
```suggestion
```

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 2]
```diff
@@ -898,6 +938,192 @@ export class CkbHandler {
...
+              }
+              return value;
+            });
+          });
+          const transaction = createTransactionFromSkeleton(signedTxSkeleton);
+          if (sigs.length < ForceBridgeCore.config.ckb.multisigScript.M) {
+            throw new Error(
+              `createBridgeCell collect signatures failed, expected:${ForceBridgeCore.config.ckb.multisigScript.M}, collected:${sigs.length}`,
+            );
+          }
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:1055]: Move the code up, check sigs.length before construct tx

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 3]
```diff
@@ -898,6 +938,192 @@ export class CkbHandler {
...
+          ).serializeJson();
+          const signedTxSkeleton = txSkeleton.update('witnesses', (witnesses) => {
+            return witnesses.map((value, index) => {
+              if (index === 0) {
+                return witness;
+              }
+              return value;
+            });
+          });
+          const transaction = createTransactionFromSkeleton(signedTxSkeleton);
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:1050]: Miss collector signature

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 4]
```diff
@@ -898,6 +938,192 @@ export class CkbHandler {
...
+      return;
+    }
+    logger.debug(`unlock for records, allTxSkeletonWithRecords`, records, allTxSkeletonWithRecords);
+    logger.debug(
+      `unlock tx txSkeletons: [ ${allTxSkeletonWithRecords
+        .map((txs) => transactionSkeletonToJSON(txs.txSkeleton))
+        .join(', ')} ]`,
+    );
+    for (const txSkeletonWithRecords of allTxSkeletonWithRecords) {
+      for (;;) {
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:1015]: Need refactor according to: https://github.com/nervosnetwork/force-bridge/blob/074c4617065236bd531508e04373b21038802a3a/offchain-modules/packages/x/src/handlers/ckb.ts#L511

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 5]
```diff
@@ -898,6 +938,192 @@ export class CkbHandler {
...
+
+  handleTodoUnlockRecords(ownerTypeHash: string, generator: CkbTxGenerator): void {
+    foreverPromise(async () => this.todoUnlockRecordsHandler(ownerTypeHash, generator), {
+      onRejectedInterval: 15000,
+      onResolvedInterval: 15000,
+      onRejected: (e: Error) => {
+        logger.error(`CKB handleTodoUnlockRecords error:${e.stack}`);
+      },
+    });
+  }
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:959]: Unused code?

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 6]
```diff
@@ -761,6 +771,36 @@ export class CkbHandler {
...
+  ): Promise<string[]> {
+    const trickyLumosConfig = {
+      PREFIX: generator.lumosConfig.PREFIX,
+      SCRIPTS: {
+        SECP256K1_BLAKE160_MULTISIG: generator.lumosConfig.SCRIPTS.SECP256K1_BLAKE160_MULTISIG,
+      },
+    };
+    txSkeleton = secp256k1Blake160Multisig.prepareSigningEntries(txSkeleton, {
+      config: trickyLumosConfig,
+    });
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:787]: No need to prepareSigningEntries again, already done in `fn generator.unlock()`

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts [hunk 7]
```diff
@@ -812,6 +818,45 @@ export class CkbTxGenerator extends CkbTxHelper {
...
   }
 
+  async unlock(records: ICkbUnlock[]): Promise<TransactionSkeletonWithUnlockRecord[]> {
+    const ethereumRecords = records.filter((r) => r.xchain === ChainType.ETH);
+    const ckbRecords = ethereumRecords.filter(
+      (r) => !r.assetIdent || r.assetIdent === '0x0000000000000000000000000000000000000000000000000000000000000000',
+    );
+    const sudtRecords = ethereumRecords.filter(
+      (r) => r.assetIdent && r.assetIdent !== '0x0000000000000000000000000000000000000000000000000000000000000000',
+    );
```
[the remark this answers is not in the document]
Comment (pygman) [reply offchain-modules/packages/x/src/ckb/tx-helper/generator.ts]: refactored, select only one kind asset that are the same as the latest one (up to 50)

File: offchain-modules/packages/x/src/handlers/ckb.ts (hunk 1 above)
Comment (pygman) [reply offchain-modules/packages/x/src/handlers/ckb.ts]: removed TransferOutSwitch currently

File: offchain-modules/packages/x/src/handlers/ckb.ts (hunk 2 above)
Comment (pygman) [reply offchain-modules/packages/x/src/handlers/ckb.ts]: fixed

File: offchain-modules/packages/x/src/handlers/ckb.ts (hunk 3 above)
Comment (pygman) [reply offchain-modules/packages/x/src/handlers/ckb.ts]: add collector signature according to the signature of update omni-lock admin cell

File: offchain-modules/packages/x/src/handlers/ckb.ts (hunk 4 above)
Comment (pygman) [reply offchain-modules/packages/x/src/handlers/ckb.ts]: refactored, cause of unlock only one kind asset, refactored according to ckbMintHandler

File: offchain-modules/packages/x/src/handlers/ckb.ts (hunk 5 above)
Comment (pygman) [reply offchain-modules/packages/x/src/handlers/ckb.ts]: removed currently

File: offchain-modules/packages/x/src/handlers/ckb.ts (hunk 6 above)
Comment (pygman) [reply offchain-modules/packages/x/src/handlers/ckb.ts]: fixed, reviewed generator unlock and understood prepareSigningEntries function

**Pull request merged**